source/common/stats/heap_stat_data.cc

```diff
@@ -2,29 +2,35 @@
 
 #include "common/common/lock_guard.h"
 #include "common/common/thread.h"
+#include "common/common/utility.h"
 
 namespace Envoy {
 namespace Stats {
 
-HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
+HeapStatData::HeapStatData(absl::string_view key) {
+  StringUtil::strlcpy(name_, key.data(), key.size() + 1);
+}
 
 HeapStatDataAllocator::HeapStatDataAllocator() {}
 
 HeapStatDataAllocator::~HeapStatDataAllocator() { ASSERT(stats_.empty()); }
 
-HeapStatData* HeapStatDataAllocator::alloc(absl::string_view name) {
+HeapStatData* HeapStatDataAllocator::alloc(absl::string_view name) noexcept {
   // Any expected truncation of name is done at the callsite. No truncation is
-  // required to use this allocator.
-  auto data = std::make_unique<HeapStatData>(name);
+  // required to use this allocator. Note that data must be freed by calling
+  // its free() method, and not by destructing or unique_ptr. So this method
```

Comment:
Up to you, but you could use CSmartPtr (search code) to use RAII around calling free().